src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -181,12 +193,104 @@ private bool TryExtractStringConcatArgs(BoundExpression lowered, out ImmutableAr
                                 }
                             }
                         }
+
+                        if ((object)method == _compilation.GetWellKnownTypeMember(WellKnownMember.System_String__Concat_2ReadOnlySpans) ||
+                            (object)method == _compilation.GetWellKnownTypeMember(WellKnownMember.System_String__Concat_3ReadOnlySpans) ||
+                            (object)method == _compilation.GetWellKnownTypeMember(WellKnownMember.System_String__Concat_4ReadOnlySpans))
+                        {
+                            // Faced a span-based string.Concat call. Since we can produce such call on the previous iterations ourselves, we need to unwrap it.
+                            // The key thing is that we need not to only extract arguments, but also unwrap them from being spans and for chars also wrap them into `ToString` calls.
+                            // Another challenge is that we may have merged user-written span-based string.Concat with additional argument previously and we need to undo this change as well
+                            // so that if at some point we exceed 3 or 4 span arguments we can undo all span-concat changes and use string.Concat(string[]) overload with the same arguments as the user provided.
+                            // We do that by tracking which argument nodes have `WasCompilerGenerated` flag. If If a consecutive span of nodes is not compiler-generated, they are arguments from original user-written span.Concat call
```

Comment:
This sounds unnecessary complicated to me and I do not think the complexity is worth it. So, I suggest either  sticking to the recommended approach (unwrapping only span concat calls that we create), or unwrapping all of them, but nothing in between. I also suggest avoiding any changes around  unwrapping string.Concat calls and unwrap all of them as, I think, we used to do.

Reply:
To clarify, I do not think we need to worry about recursively unwrapping arguments of user made concat calls. Let them be as they are - strings (they can't be spans, right?). Only top level calls should be unwrapped.

Reply:
Our primary goal is C# concatenation operator(s). For everything else, we take improvements if they come for free, but will not explicitly focus on improving other scenarios.